### PR TITLE
backend: (riscv) j reg allocator is block naive with 0 real registers

### PR DESCRIPTION
**Pull request by superlopuh**
A bit of entropy minimisation, with this PR we can allocate for loops also.

### CODE REVIEW

**Review by nazavode — APPROVED** [review]
Nice convergence of things.

File: xdsl/backend/riscv/register_allocation.py [hunk 1]
```diff
@@ -155,6 +155,97 @@ def __init__(self, limit_registers: int | None = None) -> None:
...
+                self.idx += 1
+            else:
+                reg.type = reg_type(available_regs.pop())
+
+            return True
+
+        return False
+
+    def allocate_for_loop(self, loop: riscv_scf.ForOp) -> None:
+        yield_op = loop.body.block.last_op
```
Comment (superlopuh) [inline xdsl/backend/riscv/register_allocation.py:175]: this is just copy/pasted from below

**Comment by superlopuh** [comment]
Wait till you see where the register queue comes into all this! ;)

**Review by adutilleul — APPROVED** [review]

**Comment by tobiasgrosser** [comment]
@superlopuh, I find this PR really confusing. This is not just deleting the j allocator, but it's pretty much copy-pasting the j-allocator with loop support into the block naive allocator. In which way is the block naive allocator still the block naive allocator?

**Comment by tobiasgrosser** [comment]
Let's maybe discuss this on zulip.

**Pull request merged**